torchvision/datasets/triplet.py

```diff
@@ -0,0 +1,139 @@
+
+    Args:
+        class_samples (list[list]): bins of samples, binned by class
+        num_triplets (int): number of triplets to be generated
+
+    Returns:
+        list[tuple]: triplets of the form (anchor, positive, negative)
+    """
+    triplets = []
+    class_dist = Multinomial(2, torch.Tensor([1.] * len(class_samples)))
+    swap = Bernoulli(torch.Tensor([0.5]))
+    for _ in range(num_triplets):
+        pos_cls, neg_cls = get_sampled_indices(class_dist, 2)
+        if swap.sample().item() == 1:
+            pos_cls, neg_cls = neg_cls, pos_cls
+
+        pos_samples, neg_samples = class_samples[pos_cls], class_samples[neg_cls]
+
+        pos_dist = Multinomial(2, torch.Tensor([1.] * len(pos_samples)))
+        neg_dist = Multinomial(1, torch.Tensor([1.] * len(neg_samples)))
+
+        anc_idx, pos_idx = get_sampled_indices(pos_dist, 2)
+        neg_idx = get_sampled_indices(neg_dist, 1)[0]
+        if swap.sample().item() == 1:
+            anc_idx, pos_idx = pos_idx, anc_idx
+
+        triplet = (pos_samples[anc_idx], pos_samples[pos_idx], neg_samples[neg_idx])
+        triplets.append(triplet)
+
+    return triplets
+
+
+class TripletDataset(data.IterableDataset, DatasetFolder):
+    """
+    A dataset with samples of the form (anchor, positive, negative), where anchor and
+    positive are samples of the same class, and negative is a sample of another class.
+    The dataset reads samples from a directory arranged in the following manner: ::
+
+        root/class_x/xxx.ext
+        root/class_x/xxy.ext
+        root/class_x/xxz.ext
+
+        root/class_y/123.ext
+        root/class_y/nsdf3.ext
+        root/class_y/asd932_.ext
+
+    Args:
+        root (string): Root directory path.
+        loader (callable): A function to load a sample given its path.
+        extensions (tuple[string]): A list of allowed extensions,
+        transform (callable, optional): A function/transform that takes in
+            a sample and returns a transformed version.
+            E.g, ``transforms.RandomCrop`` for images.
+        is_valid_file (callable, optional): A function that takes the path of an Image
+        file and checks if the file is a valid_file. Both extensions and is_valid_file
+        should not be passed.
+
+    Attributes:
+        samples (list[tuple]): List of (anchor, positive, negative) triplets
+    """
+        
+    def __init__(self, root, loader, num_triplets, extensions=None, transform=None, is_valid_file=None):
+        super(TripletDataset, self).__init__(root, loader,
+                                             extensions=extensions,
+                                             transform=transform,
+                                             is_valid_file=is_valid_file)
+        self.num_triplets = num_triplets
+
+        self.class_samples = get_class_samples(len(self.classes), self.samples)
+        self.samples = generate_triplets(self.class_samples, self.num_triplets)
```

Comment:
Can the samples instead be generated on-the-fly in the iterator? That way the negative samples can vary across epochs.

Reply:
Sure, I'll change that. I thought it would be a bad idea to have samples vary across epochs, which is why I generated them in the constructor. 

Would you like the anchor and positive parts of the triplet to be generated on the fly too? Or would you like those to be consistent over epochs?

Reply:
I think the negatives for sure should be random. For the positive ones, maybe have a `shuffle=` kwarg? What do you think? Also, @fmassa may have better ideas.

Reply:
After putting some thought into it, I definitely agree that negatives should be random, since there will be a lot more negatives than positives for a specific anchor. @fmassa thoughts?

I'm not sure if I quite understand what you mean by shuffle in this case. Do you mean to generate entire triplets on-the-fly or to generate `(anchor, positive)` tuples in the constructor and then shuffle these tuples for every epoch?

Reply:
Nevermind, I thought about it more and the `shuffle` thing doesn't quite make sense. I was thinking about generating all positive pairs at ctor if `shuffle=False` but I don't see its purpose...

Reply:
Ahh, I see. Let's see what @fmassa has to say about this conversation before I make any changes.

Reply:
I think the whole purpose of having the `TripletDataset` be an `IterableDataset` would be that this dataset is random by nature, and the way I see a `Dataset` should preferably not have randomness in its `__getitem__` to select which images to choose.

Thus, my thinking was to have the samples be generated on-the-fly in the iterator, for the reasons mentioned by @SsnL

Reply:
Got it. So this is what I plan on doing:

1. Change `__next__` to generate a random valid triplet on-the-fly, as opposed to iterating over `self.samples`

But then I'm not sure why we need `num_triplets` at all, since there's no reason to raise `StopIteration`. If we decide on keeping the `num_triplets` argument, I plan on doing the following:

2. After iterating over `num_triplets /  num_workers` samples for each worker process, I will raise a `StopIteration` command.

Should we keep `num_triplets`? And does this plan sound good?

Reply:
I think `num_triplets` is necessary so that we don't have an infinite dataset. Lots of pipelines assumes that we run the code over a number of "epochs", so stopping from time to time is a good thing to have.

Reply:
Alright, I have made these changes.